keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+
+## Motivation
+
+This enhancement proposal aims at improving the quality of service of
+applications by making available a new type of resource control mechanism in
+Kubernetes. Certain types of resources are inherently shared by applications,
+e.g. cache, memory bandwidth and disk I/O. While there are technologies for
+controlling how these resources are shared between applications, there is
+currently no meaningful way to support these technologies in Kubernetes. This
+proposal suggests to address the issue above in a generalized way by extending
+the Kubernetes resource model with a new type of resources, i.e. QoS-class
+resources.
+
+This KEP identifies two technologies that can immediately be enabled with
+QoS-class resources. However, these are just two examples  and the proposed
+changes are generic (and not tied to these two QoS-class resource types in any
+way), making it easier to implement new QoS-class resource types.
+
+[Intel RDT][intel-rdt] implements a class-based mechanism for controlling the
+cache and memory bandwidth QoS of applications. All processes in the same
+hardware class share a portion of cache lines and memory bandwidth. RDT
+provides a way for mitigating noisy neighbors and fulfilling SLAs. In Linux
+control happens via resctrl -- a pseudo-filesystem provided by the kernel which
+makes it virtually agnostic of the hardware architecture. The OCI runtime-spec
+has supported Intel RDT for a while already. Other hardware vendors have
+comparable technologies which use the same [resctrl interface][linux-resctrl].
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes has been
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients. Following this model also
+provides a possible framework for the future improvements, for instance
+enabling class-based network or memory type prioritization of applications.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources.
+CRI-O and containerd runtimes have support for RDT and blockio classes and they
+provide an bridge-gap user interface through special pod annotations. The goal
+is to get these types of resources first class citizens and properly supported
+in Kubernetes, providing visibility, a well-defined user interface, and
+permission controls.
+
+It seems necessary to support both container-level and pod-level QoS-class
+resources as independent concepts. Intel RDT (above) is per-container by
+design because of the hardware implementation (the control/class hiearchy is
+flat). Also, the current support for blockio is container-level only (it is not
```

Comment:
This doesn't make sense to me. Pods have cgroups and various (kernel) namespaces, right? What mechanism or architectural restriction stops us from applying the `resctl` allocation to the entire Pod cgroup?

Reply:
The resctl is flat structure. It is not like cgroup, so you have only limited (by hardware) set of classes defined, and then you can attach process to only one of those classes. In theory, we can put the whole PIDs from all cgroup subdirs for the pod to the resctl pid list, but that would require plumbing on the lower layers, including but not limiting to OCI spec, runc, runtime shims... 

And that's probably should be done as separate story: the container level cgroups/resctl manipulations are done by runtimes in accordance of OCI spec, and the whole Pod level cgroups management that is currently implemented inside kubelet. If we want to play nicely with pod level resources, we need to improve CRI protocol to have proper contract "who is doing what".

Reply:
Rather than talk about RDT here, which I think is a bit of a red herring, how about using an example of something which is shared across a Pod. Network interface is the easiest example to explain as Pods conventionally have just one, but there are some other examples like volume I/O performance class (storage attachments are at the Pod level, even if not all containers access all defined volumes).

It's OK to keep the detail about RDT shortcomings as an aside to wrap things up and show where it can be a bit tricky to work within existing restrictions. Having different RDT classes per container is plausible, whereas a per-container QoS measure on the shared netns might well not even make sense as an idea.

Reply:
I think this could possibly be resolved, see https://github.com/kubernetes/enhancements/pull/3004#discussion_r1096100072

Reply:
The wording here is unchanged AIUI - did you want to update it at all? I recommend mentioning RDT last or as an aside

> in theory, we can put the whole PIDs from all cgroup subdirs for the pod to the resctl pid list

I don't like to constrain our API design - that we may have to live with essentially forever - if the main motivation is that it's tricky to do a thing we actually _can_ do.

There are other things we can do with K8s - for example, we can launch multiple containers in a Pod that all have the same netns, or that all have the same ServiceAccount token mounted. If we can achieve those across multiple containers than I don't see that making multiple containers in a Pod share a `resctl` assignment is particularly tricky. We either control creation of the threads in that Pod or they are cloned from a thread that we did control.

We shouldn't have to worry about processes that are in the Pod but not in any of the Pod's containers, because there aren't any (there's the `pause` process; we run that in a container too).

I'm assuming that `resctl` assignments survive any `clone()` call that an unprivileged container process can make. If that doesn't hold, though, we've got a bigger problem anyway.

Reply:
`resctl` survives `clone()`. However, processes in containers and "pod process" are not the same parent, as far as I recall. In any case, initial assigning of container to particular cache/membw class is easy task. updating that, to assign it to another class is not so trivial case.